sunpy/data/data_manager/cache.py

```diff
@@ -0,0 +1,139 @@
+        redownload: `bool`
+            Whether to skip cache and redownload.
+
+        Returns
+        -------
+        `pathlib.PosixPath`
+            Path to the downloaded file.
+        """
+        if isinstance(urls, str):
+            urls = [urls]
+        # Program flow
+        # 1. If redownload: Don't check cache, don't put in cache. Download and return file path
+        # 2. If not redownload: Check cache,
+        #    i. If present in cache:
+        #        - If cache expired, remove entry from cache, download and add to cache
+        #        - If cache not expired, return path
+        #    ii. If not download, store in cache and return path
+        if not redownload:
+            details = None
+            for url in urls:
+                details = self._get_by_url(url)
+                if details:
+                    break
+            if details:
+                if self._expiry and \
+                   datetime.now() - datetime.fromisoformat(details['time']) > self._expiry:
+                    os.remove(details['file_path'])
+                    self._storage.delete_by_key('url', details['url'])
+                else:
+                    return Path(details['file_path'])
+
+        file_path, file_hash, url = self._download_and_hash(urls)
+
+        if not redownload:
+            self._storage.store({
+                'file_hash': file_hash,
+                'file_path': str(file_path),
+                'url': url,
+                'time': datetime.now().isoformat(),
+            })
+        return file_path
+
+    def get_by_hash(self, sha_hash):
+        """
+        Returns the details which is matched by hash if present in cache.
+
+        Parameters
+        ----------
+        sha_hash: `str`
+            SHA-1 hash of the file.
+        """
+        details = self._storage.find_by_key('file_hash', sha_hash)
+        return details
+
+    def _get_by_url(self, url):
+        """
+        Returns the details which is matched by url if present in cache.
+
+        Parameters
+        ----------
+        url: `str`
+            URL of the file.
+        """
+        details = self._storage.find_by_key('url', url)
+        return details
+
+    def _download_and_hash(self, urls):
+        """
+        Downloads the file and returns the path, hash and url it used to download.
+
+        Parameters
+        ----------
+        urls: `list`
+            List of urls.
+
+        Returns
+        -------
+        `str`, `str`, `str`
+            Path, hash and URL of the file.
+        """
+        def download(url):
+            path = self._cache_dir / get_filename(urlopen(url), url)
+            path = replacement_filename(path)
+            self._downloader.download(url, path)
+            shahash = hash_file(path)
+            return path, shahash, url
+
+        for url in urls:
+            try:
+                return download(url)
+            except Exception as e:
+                warn(e)
```

Comment:
We don't need to add this to the sunpy logger?

Reply:
Use `SunpyUserWarning` here